File: leetcode/1249-minimum-remove-to-make-valid-parentheses.py
# https://leetcode.com/problems/minimum-remove-to-make-valid-parentheses/

class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        s = list(s)
        stack = []
        for i in range(len(s)):
            if stack and s[i] == ')' and stack[-1]['parenthese'] == '(':
                stack.pop()
            elif s[i] == '(' or s[i] == ')':
                stack.append({'parenthese': s[i], 'index': i})
        
        for invalid_parenthese in stack:
            s[invalid_parenthese['index']] = ''
        
        return "".join(s)

File: leetcode/1249-minimum-remove-to-make-valid-parentheses.js
// https://leetcode.com/problems/minimum-remove-to-make-valid-parentheses/

/**
 * @param {string} s
 * @return {string}
 */
var minRemoveToMakeValid = function(s) {
  s = s.split('');
  const stack = []; // stack of invalid parentheses with their indicies...
  for (let i = 0; i < s.length; i += 1) {
    if (stack.length > 0 && s[i] === ')' && stack[stack.length - 1].parenthese === '(') {
      stack.pop();
    } else if (s[i] === '(' || s[i] === ')') {
      stack.push({parenthese: s[i], index: i});
    }
  }

  for(const { index } of stack) {
    s[index] = null;
  }

  return s.join('');
};

